server/src/main/java/org/apache/lucene/search/grouping/CollapsingTopDocsCollector.java

```diff
@@ -92,15 +109,21 @@ public ScoreMode scoreMode() {
     }
 
     @Override
-    public void setScorer(Scorable scorer) throws IOException {
-        super.setScorer(scorer);
-        this.scorer = scorer;
+    protected void doSetNextReader(LeafReaderContext readerContext) throws IOException {
+        leafComparator = comparator.getLeafComparator(readerContext);
```

Comment:
I found it a little simpler to reason about if we new `leafComparator` was always populated and non-null. I also didn't think this would be too expensive since it's not called often?